### PR TITLE
Bug 1054303 - Use open search plugins for search engines

### PR DESCRIPTION
**Pull request by leibovic**
Still a WIP, but I wanted to open a pull request to share my progress.

### CODE REVIEW

File: app/src/androidTest/java/org/mozilla/search/test/SearchEngineTest.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    public void testIsResultsPage() {
+        assertFalse(engine.isSearchResultsPage("http://example.com"));
+        assertTrue(engine.isSearchResultsPage("http://www.google.com/search"));
+    }
+
+    public void testResultsUri() {
+        final String expected = "http://www.google.com/search?q=query&ie=utf-8&oe=utf-8&aq=t";
+        assertEquals(expected, engine.resultsUriForQuery("query"));
```
Comment (EricEdens) [inline app/src/androidTest/java/org/mozilla/search/test/SearchEngineTest.java]: If url params change order, this could create a false negative. I think the .equals method of Uri could help to insulate from param order:

http://developer.android.com/reference/android/net/Uri.html#equals%28java.lang.Object%29

File: app/src/main/java/org/mozilla/search/providers/SearchEngine.java [hunk 2]
```diff
@@ -21,82 +37,167 @@
...
 
     /**
-     * Retrieve a JS snippet, in bookmarklet style, that can be used
-     * to modify the results page.
+     *
+     * @param in InputStream of open search plugin XML
      */
-    public String getInjectableJs() {
-        return String.format(STYLE_INJECTION_SCRIPT, getInjectableCss());
+    public SearchEngine(String identifier, InputStream in) throws IOException, XmlPullParserException {
```
Comment (EricEdens) [inline app/src/main/java/org/mozilla/search/providers/SearchEngine.java:59]: Maybe move the parsing / construction into another class? Maybe even an inner class? I think the benefit would be keeping the SearchEngine class short, so it's easy to see what fields / method are available within the UI.

File: app/src/main/java/org/mozilla/search/providers/SearchEngineManager.java [hunk 3]
```diff
@@ -63,73 +59,88 @@ public void destroy() {
...
+     * @param identifier
+     * @return
      */
-    private void updateEngine(String engineName) {
+    private SearchEngine createEngine(String identifier) {
+        InputStream in = getEngineFromJar(identifier);
 
-        if (TextUtils.isEmpty(engineName)) {
-            engineName = Constants.DEFAULT_SEARCH_ENGINE;
+        // Fallback for standalone search activity
```
Comment (EricEdens) [inline app/src/main/java/org/mozilla/search/providers/SearchEngineManager.java]: Nice :)

File: app/src/main/java/org/mozilla/search/providers/SearchEngineManager.java [hunk 4]
```diff
@@ -63,73 +59,88 @@ public void destroy() {
...
      */
     private void getEngineFromPrefs(final SearchEngineCallback callback) {
-        final AsyncTask<Void, Void, String> task = new AsyncTask<Void, Void, String>() {
+        final AsyncTask<Void, Void, SearchEngine> task = new AsyncTask<Void, Void, SearchEngine>() {
             @Override
-            protected String doInBackground(Void... params) {
-                return GeckoSharedPrefs.forApp(context).getString(SearchPreferenceActivity.PREF_SEARCH_ENGINE_KEY, null);
+            protected SearchEngine doInBackground(Void... params) {
+                final String identifier = GeckoSharedPrefs.forApp(context).
+                        getString(SearchPreferenceActivity.PREF_SEARCH_ENGINE_KEY, Constants.DEFAULT_SEARCH_ENGINE).
```
Comment (EricEdens) [inline app/src/main/java/org/mozilla/search/providers/SearchEngineManager.java]: I've usually seen the wrapped period on the next line. Do we put it on the previous in fennec?

File: app/src/androidTest/java/org/mozilla/search/test/SearchEngineTest.java (hunk 1 above)
Comment (leibovic) [inline app/src/androidTest/java/org/mozilla/search/test/SearchEngineTest.java]: Good idea. I realized these tests were pretty fragile, but at least they would be fragile in the false-positive failure direction ;)

File: app/src/main/java/org/mozilla/search/providers/SearchEngineManager.java [hunk 5]
```diff
@@ -10,24 +10,23 @@
...
 import org.mozilla.search.Constants;
 import org.mozilla.search.SearchPreferenceActivity;
+import org.xmlpull.v1.XmlPullParserException;
+
+import java.io.IOException;
+import java.io.InputStream;
 
 public class SearchEngineManager implements SharedPreferences.OnSharedPreferenceChangeListener {
-    private static final String LOG_TAG = "SearchEngineManager";
+    private final static String LOG_TAG = "SearchEngineManager";
```
Comment (thebnich) [inline app/src/main/java/org/mozilla/search/providers/SearchEngineManager.java:24]: Nit: Unnecessary change

Comment (leibovic) [inline app/src/main/java/org/mozilla/search/providers/SearchEngineManager.java:24]: Oops, I must have done that in different iterations that removed/added this line.
